File: api/src/main/java/com/example/sms/domain/model/master/employee/Employee.java
package com.example.sms.domain.model.master.employee;

import com.example.sms.domain.model.master.department.Department;
import com.example.sms.domain.model.master.department.DepartmentId;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Value;

import java.time.LocalDateTime;

/**
 * 社員
 */
@Value
@Getter
@AllArgsConstructor
@NoArgsConstructor(force = true)
public class Employee {
    EmployeeCode empCode;       // 社員コード
    EmployeeName empName;       // 社員名
    String loginPassword; // パスワード
    PhoneNumber tel;           // 電話番号
    FaxNumber fax;           // fax番号
    DepartmentId departmentId; // 部門ID
    String occuCode;      // 職種コード
    String approvalCode;  // 承認権限コード
    Department department; // 部門

    public static Employee of(String empCode, String name, String kana, String password, String tel, String fax, String deptCode, LocalDateTime startDate, String occuCode, String approvalCode) {
        EmployeeCode employeeCode = EmployeeCode.of(empCode);
        DepartmentId departmentId = DepartmentId.of(deptCode, startDate);
        EmployeeName employeeName = EmployeeName.of(name, kana);
        PhoneNumber phoneNumber = PhoneNumber.of(tel);
        FaxNumber faxNumber = FaxNumber.of(fax);
        return new Employee(employeeCode, employeeName, password, phoneNumber, faxNumber, departmentId, occuCode, approvalCode, null);
    }

    public static Employee of(String empCode, String name, String kana, String password, String tel, String fax, String deptCode, LocalDateTime startDate, String occuCode, String approvalCode, Department department) {
        EmployeeCode employeeCode = EmployeeCode.of(empCode);
        DepartmentId departmentId = DepartmentId.of(deptCode, startDate);
        EmployeeName employeeName = EmployeeName.of(name, kana);
        PhoneNumber phoneNumber = PhoneNumber.of(tel);
        FaxNumber faxNumber = FaxNumber.of(fax);
        return new Employee(employeeCode, employeeName, password, phoneNumber, faxNumber, departmentId, occuCode, approvalCode, department);
    }
}

File: app/frontend/src/components/master/hooks/employee.ts
import {useState} from "react";
import {
    EmployeeType,
    LowerType,
    SlitYnType
} from "../../../models";
import {EmployeeService, EmployeeServiceType} from "../../../services/employee.ts";
import {PageNationType} from "../../../views/application/PageNation.tsx";
import {useFetchEntities} from "../../application/hooks.ts";

export const useEmployee = () => {
    const initialEmployee: EmployeeType = {
        empCode: {value: ""},
        empName: {
            firstName: "",
            lastName: "",
            firstNameKana: "",
            lastNameKana: ""
        },
        loginPassword: "",
        tel: {
            value: "",
            areaCode: "",
            localExchange: "",
            subscriberNumber: ""
        },
        fax: {
            value: "",
            areaCode: "",
            localExchange: "",
            subscriberNumber: ""
        },
        occuCode: {
            value: ""
        },
        approvalCode: "",
        department: {
            departmentId: {deptCode: {value: ""}, departmentStartDate: {value: ""}},
            endDate: {value: ""},
            departmentName: "",
            layer: 0,
            path: {value: ""},
            lowerType: LowerType.NO,
            slitYn: SlitYnType.NO,
            employees: [],
            checked: false
        },
        user: {
            userId: {value: ""},
            name: {
                firstName: "",
                lastName: ""
            },
            password: {
                value: ""
            },
            roleName: ""
        },
        addFlag: false,
        deleteFlag: false,
        checked: false
    };

    const [employees, setEmployees] = useState<EmployeeType[]>([]);
    const [newEmployee, setNewEmployee] = useState<EmployeeType>(initialEmployee);
    const [searchEmployeeCode, setSearchEmployeeCode] = useState<string>("");

    const employeeService = EmployeeService();

    return {
        initialEmployee,
        employees,
        newEmployee,
        setNewEmployee,
        setEmployees,
        searchEmployeeCode,
        setSearchEmployeeCode,
        employeeService,
    };
}

export const useFetchEmployees = (
    setLoading: (loading: boolean) => void,
    setList: (list: EmployeeType[]) => void,
    setPageNation: (pageNation: PageNationType) => void,
    setError: (error: string) => void,
    showErrorMessage: (message: string, callback: (error: string) => void) => void,
    service: EmployeeServiceType
) => useFetchEntities(setLoading, setList, setPageNation, setError, showErrorMessage, service, "社員情報の取得に失敗しました:");